chainer/functions/connection/depthwise_convolution_2d.py

```diff
@@ -0,0 +1,183 @@
+
+        if xp is numpy:
+            gx = conv.col2im_cpu(gcol, self.sy, self.sx,
+                                 self.ph, self.pw, h, w)
+        else:
+            gx = conv.col2im_gpu(gcol, self.sy, self.sx,
+                                 self.ph, self.pw, h, w)
+
+        if b is None:
+            return gx, gW
+        else:
+            gy = xp.rollaxis(gy, 1, 4)
+            gb = gy.sum(axis=(0, 1, 2))
+            return gx, gW, gb
+
+
+def depthwise_convolution_2d(x, W, b=None, stride=1, pad=0):
+    """Two-dimensional depthwise convolution function.
+
+    This is an implementation of two-dimensional depthwise convolution.
+    It takes three variables: the input image ``x``, the filter weight ``W``,
+    and the bias vector ``b``.
+
+    Notation: here is a notation for dimensionalities.
+
+    - :math:`n` is the batch size.
+    - :math:`c_I` is the number of the input.
+    - :math:`c_M` is the number of the channel multiplier.
+    - :math:`h` and :math:`w` are the height and width of the input image,
+      respectively.
+    - :math:`h_O` and :math:`w_O` are the height and width of the output image,
+      respectively.
+    - :math:`k_H` and :math:`k_W` are the height and width of the filters,
+      respectively.
+
+    Args:
+        x (chainer.Variable or :class:`numpy.ndarray` or cupy.ndarray):
+            Input variable of shape :math:`(n, c_I, h, w)`.
+        W (~chainer.Variable): Weight variable of shape
+            :math:`(c_M, c_I, k_H, k_W)`.
+        b (~chainer.Variable):
+            Bias variable of length :math:`c_M * c_I` (optional).
+        stride (int or pair of ints): Stride of filter applications.
+            ``stride=s`` and ``stride=(s, s)`` are equivalent.
+        pad (int or pair of ints): Spatial padding width for input arrays.
+            ``pad=p`` and ``pad=(p, p)`` are equivalent.
+
+
+    Returns:
+        ~chainer.Variable:
+            Output variable. The shape is :math:`(n, c_I * c_M, h_O, w_O)`.
+
+    Like ``Convolution2D``, ``DepthwiseConvolution2D`` function computes
+    correlations between filters and patches of size :math:`(k_H, k_W)` in
+    ``x``.
+    But unlike ``Convolution2D``, ``DepthwiseConvolution2D`` not adds up
+    between input channels of filters but concatenates.
+    For that reason, the shape of outputs of depthwise convolution are
+    :math:`(n, c_I * c_M, h_O, w_O)`, :math:`c_M` called channel_multiplier.
+
+    :math:`(h_O, w_O)` is determined by the equivalent equation with
+    ``Convolution2D``.
+
+    If the bias vector is given, then it is added to all spatial locations of
+    the output of convolution.
+
+    See: `L. Sifre. Rigid-motion scattering for image classification\
+          <http://www.di.ens.fr/data/publications/papers/phd_sifre.pdf>`_
+
+    .. seealso:: :class:`~chainer.links.DepthwiseConvolution2D`
+
+    .. admonition:: Example
+
+        >>> x = np.random.uniform(0, 1, (2, 3, 4, 7)).astype('f')
```

Comment:
Is `astype('f')` necessary? `np.random.uniform` returns `numpy.float64` and it should pass `check_type_forward` of this function.